### ISSUE TITLE
onDelete cascade for paranoid tables

### ISSUE THREAD

**Issue by lao** [sequelize/sequelize#2586]
I have a relationship between two paranoid tables (User and Email) and I have the onDelete: 'cascade' option on the relationship ( Email.belongsTo(User, onDelete:'cascade') ). 

The problem is that when I delete the user his email is **not** being deleted by cascade.

Is this a bug? Or this is how it is supposed to work?

I am using **sequelize 2.0.0-rc2** on a **postgres** database. 

Thanks.

PS.: Please take a look at the code I used for test:

``` javascript
var Sequelize = require('sequelize')
    , sequelize = new Sequelize('test', 'test', 'test', {dialect: 'postgres'});

var User = sequelize.define('User', {
    username: Sequelize.STRING,
    birthday: Sequelize.DATE
}, {paranoid: true}); //Without paranoid the code works fine

var Email = sequelize.define('Email', {
    email: Sequelize.STRING,
    primary: Sequelize.BOOLEAN
}, {paranoid: true}); //Without paranoid the code works fine

Email.belongsTo(User, {onDelete: 'cascade'});

sequelize.sync().success(function () {
    var userCreatedInstance = null;

    var deletedUserId = null;
    var cascadeDeletedEmailId = null;

    User
        .create({
            username: 'user1',
            birthday: new Date(1986, 06, 28)
        })
        .then(function (_user) {
            userCreatedInstance = _user;
            deletedUserId = userCreatedInstance.id;
            return Email.create({email: 'email1@test.com', primary: true, UserId: userCreatedInstance.id});
        })
        .then(function (createdEmail) {
            cascadeDeletedEmailId = createdEmail.id;
            return userCreatedInstance.destroy();
        })
        .then(function () {
            return User.find(deletedUserId);
        })
        .then(function (foundUser) {
            if(foundUser){
                throw new Error("Shouldn't find the user with this id");
            }
            return Email.find(cascadeDeletedEmailId);
        })
        .then(function (foundEmail){
            if(foundEmail){
                console.log(foundEmail.values);
                throw new Error("Shouldn't find the email with this id");
            }
        });
});
```

**Comment by clivend** [comment]
+1

**Comment by strobmax** [comment]
+1

**Comment by elsasquash** [comment]
+1

**Comment by marcoviz** [comment]
+1

**Comment by felixfbecker** [comment]
![image](https://cloud.githubusercontent.com/assets/10532611/17862296/f0aa3176-6894-11e6-81fa-558d4c09cf93.png)

**Comment by kossel** [comment]
+1

**Comment by diosney** [comment]
Any updates on this? Is planned to be included in v4?

**Comment by felixfbecker** [comment]
@diosney if it was, it would have the v4 milestone 😉 
Nobody of the team is working on it but if someone wants to implement it we can include it :)

**Comment by diosney** [comment]
Jajaja! You made my day :smile: That was an indirect? If so I'm willingly to do it with the correct guidance :wink:

**Comment by felixfbecker** [comment]
Almost all of Sequelize features come from user contributions 😉 
I am the wrong maintainer to guide you as I didn't write any code related to this - maybe you should do a git blame on the association files and ping the author. You can also join us on Slack.

**Comment by ajmueller** [comment]
@lao what solution did you eventually come up with?  I need the same functionality on an app I'm building, so I'm just using managed transactions to do this.

**Comment by richardpringle** [comment]
+1for the feature

**Comment by lao** [comment]
@ajmueller unfortunately we didn't work on a generic solution. I think we ended up using managed transactions as well. But I am not sure. It was a few years ago and I don't work on the project anymore. =/

**Comment by ajmueller** [comment]
@lao thanks for the reply.  I actually ended up not changing to not use the paranoid feature for deletes.  Instead I built a common interface whereby users must manually type "DELETE" into a text input in a modal if they want to delete something.

**Comment by trjstewart** [comment]
Anyone made any progress on this? Considering spending some time on it but not 100% sure where to start.

**Comment by alexey2baranov** [comment]
Hello!
Why this issue closed? Is it implemented?

**Comment by blankstar85** [comment]
Would be nice if the docs would indicate that the onDelete needs to be in the table creation. Sat here for a hour trying to figure out why the cascade wasn't working. I don't use the sync as I hand write the migrations to have better control of the DB. Thanks for the info on this issue. Glad i was able to find it :D

**Comment by furqanaziz** [comment]
Desperately waiting for this. For now I had to handle cascade delete manually in model hooks.

**Comment by fareshan** [comment]
+1 for the feature

**Comment by olee2002** [comment]
Desperately waiting for this too!

**Comment by aminalali8** [comment]
I think this is it. It's the first and last time I will use Sequelize so many essential features and the documentation is a cry me a river rhyme

**Comment by R-iskey** [comment]
Any updates ?
==========

**Comment by kyleguate** [comment]
# Purpose
This post is targeted to experts/maintainers of this library.  It contains my potential approach to adding this feature.  

I currently use this library at work for some production apps so I have some experience with it. 
 However I've looked at the `model.js` code for ~20 minutes so I am definitely missing a lot of context and nuance which I hope you can provide. 

I want your feedback so that I can find out how far off track I might be and if there are a lot of consequences/edge cases I haven't considered.  Thanks!

## Instance Destroy
### Step 1
Add a configuration option to the model called something like `paranoidCascade` that would accept true/false

### Step 2
In the instance `destroy` method add another `then` handler after the `afterDestroy` hook is executed that checks for `paranoid === true` and `paranoidCascade === true` and `force === false`

### Step 3
If the ^^ check evaluates to false then just `return Promise.resolve()`.

If it evaluates to true then invoke a function/method (or series of functions.methods) that
1. Finds out which tables/models are referenced in foreign key constraints
1. Hydrates each of those models
1. Invokes `destroy` for each of those models (passing in the same options as passed in the original model).  This in effect would recursively `destroy` all rows in FK referenced tables/models until they were all _soft_ deleted

### Step 4
Because each recursive invocation of the `destroy` takes care of itself and its hooks I don't think there is anything left to do...thoughts?

## Instance Restore
Similar to `Instance Destroy` but calling the `restore` method on all FK referenced tables/models

## Static
For the `static destroy` and `static restore` we'll want to do something similar, but we'll need the instances that get hydrated as they do in `options.individualHooks` check on line 2544 in `model.js`.

We've got a couple of different implementation options there and I'm open to any suggestions.

**Comment by tsurumure** [comment]
agree with @kyleguate

**Comment by Fryuni** [comment]
There is this text on the docs [here](http://docs.sequelizejs.com/manual/tutorial/hooks.html#associations):
> However, adding `hooks: true` explicitly tells Sequelize that optimization is not of your concern and will perform a SELECT on the associated objects and destroy each instance one by one in order to be able to call the hooks with the right parameters.

Doesn't that solve exctly this problem?

**Comment by Fryuni** [comment]
Just looked at the source for that and it only works when deleting, not restoring, so that would have to be implemented if wanted.  
IMHO the restore cascade should not be implemented. This way it behaves the same as deleting an entry with children cascades and creating it again directly on the DB.

**Comment by collinstowell** [comment]
Would love if we could get cascading deletes/updates in place for soft-delete / paranoid-enabled tables. Any updates/thoughts? I appreciate what you guys do.

**Comment by tommybananas** [comment]
This should be fairly trivial to implement at the application level if you delete with a transaction and then get reference to the transaction in the delete hook (`options.transaction`) to handle your own cascading. After thinking about it I agree this feature is kind of a reach for an ORM.

**Comment by papb** [comment]
@kyleguate I like your ideas, would you be willing to submit a PR?

**Comment by JayAndCatchFire** [comment]
If this is worthy of a separate ticket, let me know and I'll go open one.  I'm not 💯 about whether what I'm about to describe is the same as what's described in this ticket as this ticket has been around for 5 years.

I'm having an issue where with cascading deletion in cases where the child model is set to paranoid but the parent model is not.

Here's the scenario I've got:

```javascript
const Book = sequelize.define('Book', {
  // columns
}, {
   paranoid: true,
   classMethods: {
     associate: function(models) {
       Book.belongsTo(models.Author);
    },
  },
});

const Author = sequelize.define('Author', {
  // columns
}, {
  classMethods: {
    associate: function(models) {
      Author.Books = Module.hasMany(models.Book, {
        hooks: true,
        onDelete: 'CASCADE',
      });
    },
});
```

Note that the Book model is set to paranoid but the Author model is not.  If they were both set up in paranoid mode and I call `.destroy()` on an instance of Author, both the Author instance and any related instances of Book are correctly soft-deleted.  However, in my example above, Book is set to paranoid mode while Author is not.  In that case, if I call `.destroy()` on an instance of Author, it is hard-deleted from the database and all the related Book instances are also hard-deleted in spite of the Book model being set to paranoid mode.

Is that a known limitation/expected behavior, or is that a bug?  I can see it both ways.  On the one hand, it was an unexpected result from my point of view when I ran into this, but it also kind of makes sense as any relationship is now broken as those Books' foreign keys would point to non-existent entities in the Authors table.  (That's assuming foreign key constraints were enforced.  They are not enforced on the database level in my setup.  Also, FWIW, I'm using Postgres.)

**Comment by papb** [comment]
@JayAndCatchFire Please open a new issue (and link to this one as a related issue), I think it's better.

**Comment by gagoole777** [comment]
is there a solution for this issue? Thanks

**Comment by DRKSLV** [comment]
i think if you want paranoid delete/restore cascades to work on sequelize level you have to use model hooks

**Comment by Fryuni** [comment]
@gagoole777 have you tried what I pointed out from the docs?

**Comment by gagoole777** [comment]
@Fryuni Sorry for late answer. I am thankful to you, could you give me an example of a hook

**Comment by Fryuni** [comment]
I just tested and I was wrong. Cascading on the sequelize side is implemented when the `hooks: true` is set on the `hasMany` relation, it deletes one by one in this case.
The problem is that it is implemented on the [`QueryInterface`](https://github.com/sequelize/sequelize/blob/1e6f9af1bad84da395d6aa634f9e35fe4c5a6a2c/lib/query-interface.js#L1070), not on the `Model`. And the paranoid change to save instead of destroy is at the [`Model` level](https://github.com/sequelize/sequelize/blob/1e6f9af1bad84da395d6aa634f9e35fe4c5a6a2c/lib/model.js#L4172), calling the `QueryInterface.save`.

Seems to be easily solvable by just moving that cascading logic one layer up, or duplicating since its small in case there is something internal depending on it being where it is now.

**Comment by gagoole777** [comment]
@Fryuni Thanks for your help, I will try to implement as you have said and will let you know!

**Comment by mustyzod** [comment]
what is did to make this work is,
i added a afterDestroy hook on the parent table(base of the relatioship),
i made it get the model asssociated with with and called the destroy method on it. 
Easy as that!
I believe we could make it work this way till the maintainers give us another means

`
// User Model
const User = sequelize.define('users', {....},
{
        tableName: 'users',
        deletedAt: 'deletedAt',
        paranoid: true,
        timestamps: true,
        hooks: {
            afterDestroy: function (instance, options) {
                instance.getProduct().then(product=> product.destroy()); // Softdelete on product table
                console.log('after destroy: destroyed');
            }
        }
    }
);

// Product Model
const Product = sequelize.define('products', {....},
{
        tableName: 'products',
        deletedAt: 'deletedAt',
        paranoid: true,
        timestamps: true,
    }
)

// User relationship with product
User.hasOne(Product,
    {
        as: 'Product',
        onDelete: 'CASCADE',
        hooks: true,
        foreignKey: "userID"
    })
Product.belongsTo(User, {
    as: 'User',
    foreignKey: "userID"
});
`

**Comment by hemantreddy** [comment]
@mustyzod did this implementation of hooks and cascade work for you?

**Comment by mustyzod** [comment]
@hemantreddy yes it does worked perfectly, i made a post on it here [on Medium](https://medium.com/@mustyzod/sequelize-using-softdelete-restore-on-table-relationship-with-paranoid-2d9ac4baa2f0)

**Comment by Fryuni** [comment]
I had a wildly crazy idea here, add an association hook that automatically add this cascading hooks. Sadly associations ignore hooks coming from the Model (and therefore Sequelize) [here](https://github.com/sequelize/sequelize/blob/c013246bdc9ea388af6a393fd1f3f5646a4143b3/lib/associations/mixin.js#L27).

Second idea is to make a wrapper around the association methods. Someone could publish that while it is not in sequelize itself.
Something like this (very crude and done here on the comments, just an idea):
```javascript
const innerFunctions = {
  hasMany: {has: true, innerFunction: Model.hasMany},
  hasOne: {has: true, innerFunction: Model.hasOne},
  belongsTo: {has: false, innerFunction: Model.belongsTo},
};
for (const [name, {has, innerFunction}] of Object.entries(innerFunctions)) {
  Object.defineProperty(Model, name, {
    value(target, ...otherArguments) {
      this.options.hooks = this.options.hooks || {};
      const previousHook = this.options.hooks.afterDestroy;
      const association = innerFunction.call(this, target, ...otherArguments);

      (has ? this : target).options.hooks.afterDestroy = function(instance, options) {
        if (previousHook) previousHook.call(this, instance, options);
        instance[association.accessors.get]().then(result => {
          if (Array.isArray(result)) {
            result.forEach(resultInstance => resultInstance.destroy());
          } else {
            result.destroy();
          }
        });
      }

      return association;
    },
  });
}
```

**Comment by mi-mazouz** [comment]
any update on this issue?

**Comment by hutsi** [comment]
+1

**Comment by subashCognitive** [comment]
+1

**Comment by subashCognitive** [comment]
+200

**Comment by subashCognitive** [comment]
+1000000000

**Comment by renancsilva86** [comment]
Since 2014 waiting for this feature!

**Comment by mustyzod** [comment]
> Since 2014 waiting for this feature!

Well...i just did my part. Enjoy!

**Comment by lukasgjetting** [comment]
For anyone else looking for an easy solution, the following seems to work just fine:

```js
UserGroup.addHook('beforeDestroy', async (group) => {
	await group.setUsers([]);
});
```

If you need to do the same when users are deleted, you would need to set the same hook on the `User` model.

Our model associations are defined as follows:

```js
UserGroup.belongsToMany(models.User, {
	through: 'UsersInUserGroups',
	foreignKey: 'userGroupId',
	otherKey: 'userId',
});

User.belongsToMany(models.UserGroup, {
	through: 'UsersInUserGroups',
	as: 'groups',
	foreignKey: 'userId',
	otherKey: 'userGroupId',
});
```

**Comment by aldasmallthings** [comment]
i'm surprised this is still not a thing. HOW has this ticket been open for 7 years? I'm starting to get a lil disappointed, maintainers...

**Comment by tommybananas** [comment]
It's been open for 7 years because scalable applications don't handle cascades through an ORM layer.

**Comment by Shahaed** [comment]
> It's been open for 7 years because scalable applications don't handle cascades through an ORM layer.

@tommybananas What do you mean by this? Are you saying scalable applications don't use an ORM layer? Then they wouldn't be using Sequalize. Or they do but have some of their logic in SQL and others in Javascript? Cause that introduces another scalability issue.

**Comment by hayk-saryan** [comment]
My take on this is (maybe wrong):

We should completely ditch the usage of `onDelete: 'cascade'`, because it's not optimal for tables with large number of rows anyway. 

Instead, we could do the deletions manually: 
1. add an `afterDestroy` and `afterBulkDestroy` hooks in the model, 
2. wrap all of the destroy queries of the referenced tables (in correct order) in a single transaction 
3. and run it inside the callback of the hooks.

This approach helps as perform deletion logic using bulk operations therefore making the operation faster and lighter.

Correct me if I am wrong, better ideas and comments are welcome.

**Comment by nicoabie** [comment]
I had a new years idea and created this package to try a solution for this issue using triggers.
I believe it is a good idea as entity hooks shouldn't know (in my opinion) about other entities.
And It is also possible that SQL engines use triggers behind the scenes or some similar mechanism to achieve the cascades.

https://www.npmjs.com/package/sequelize-paranoid-cascade

It is in development but I'd like to know your thoughts.

**Comment by hayksaryan** [comment]
@nicoabie Of course, using cascading deletes is a matter of personal choice, and having a solution to keep using them is great (although I am a bit worried about performance of your solution on large datasets).

But I'd like to bring the focus on a bigger question:
**Should we delete any data from the database at all?** 

- Storage is getting cheaper and cheaper.
- A lot of times we may physically delete the data, but later a business requirement comes in which assumes we kept the data. Say, a need to generate reports about how may users deleted their account.
- There are risks with cascading deletes, e.g. you might unexpectedly delete some data that shouldn't have been deleted.
- Cascading deletes may have bad performance on large datasets.

So, my preference is to use soft deletes and handle deletion of related entites manually. Implementation of this may differ. 

I am currently using the repository pattern and update all related/child entities in root/parent entitie's repository when the root/parent entity is deleted. 

This way you don't lose database integrity, you are more flexible and ready for future business requirements.

As a bonus, your database can tell you how the data has been created, transformed, and deleted at any point of time. You never lose information.


There are also drawbacks, like more storage required and more manual work, but you are also always in control, and this is, imho, very useful when working on enterprise solutions.

But, as always, it depends 🙂

**Comment by ephys** [comment]
>  Should we delete any data from the database at all?

If you're from the European Union or have customers from the European Union, the right of erasure means you'll have to hard delete some data, not just soft delete. (disclaimer: not a lawyer).

**Comment by haykerman** [comment]
@nicoabie Of course, using cascading deletes is a matter of personal choice, and having a solution to keep using them is great (although I am a bit worried about performance of your solution on large datasets).

But I'd like to bring the focus on a bigger question:
**Should we delete any data from the database at all?** 

- Storage is getting cheaper and cheaper.
- A lot of times we may physically delete the data, but later a business requirement comes in which assumes we kept the data. Say, a need to generate reports about how many users deleted their account.
- There are risks with cascading deletes, e.g. you might unexpectedly delete some data that shouldn't have been deleted.
- Cascading deletes may have bad performance on large datasets.

So, my preference is to use soft deletes and handle deletion of related entites manually. Implementation of this may differ. 

I am currently using the repository pattern and update all related/child entities in root/parent entitie's repository when the root/parent entity is deleted. 

This way you don't lose database integrity, you are more flexible and ready for future business requirements.

As a bonus, your database can tell you how the data has been created, transformed, and deleted at any point of time. You never lose information.


There are also drawbacks, like more storage required and more manual work, but you are also always in control, and this is, imho, very useful when working on enterprise solutions.

But, as always, it depends 🙂

**Comment by haykerman** [comment]
@ephys Agreed. That kind of data can be an exception and may be handled separately.

**Comment by nicoabie** [comment]
Here we are talking for a solution to the unsupported cascade deletes in paranoid mode (meaning setting the deletedAt field)

One solution proposed was to use hooks which IMHO will be getting more responsibility than they should.

The proposed approach is to automatically creating triggers when running migrations to emulate the cascade in paranoid mode.

Is there any other way to do this that isn't cumbersome or error prone?

**Comment by hayksaryan** [comment]
> Here we are talking for a solution to the unsupported cascade deletes in paranoid mode (meaning setting the deletedAt field)
> 
> One solution proposed was to use hooks which IMHO will be getting more responsibility than they should.
> 
> The proposed approach is to automatically creating triggers when running migrations to emulate the cascade in paranoid mode.
> 
> Is there any other way to do this that isn't cumbersome or error prone?

Well, my solution is to not have a solution at all for this specific problem. Because deleting related data might be out of the scope of the ORM library,  considering it being an application's own architectural problem.

**Comment by haykerman** [comment]
Well, my solution is to not have a solution at all for this specific problem. Because deleting related data might be out of the scope of the ORM library,  considering it being an application's own architectural problem.

**Comment by nicoabie** [comment]
@haykerman I truly do not understand what you are saying. "my solution is to not have a solution at all for this specific problem" the whole point of this thread is to find a solution for this problem.
"Because deleting related data might be out of the scope of the responsibilities of an ORM library" we would not be deleting any data because it is paranoid mode (soft delete) and it won't be responsibility of the ORM but the DB as if we were using the real cascade.

 No offense or disrespect but have you read this thread?

**Comment by haykerman** [comment]
I have read the thread and I am well aware of the problem, as I've been investigating it for more than a month now.

What I am proposing is implementing deletion logic for related tables in application code, not on database level, because it's business logic. It's neither ORM's nor DB's responsibility to implement/contain this kind of business logic (imo, though this is debated). It's preferable to keep the DB as "dumb" as possible and treat it as a data store (only with some constraints). 

I am aware that some people implement the solution on DB level (through triggers or stored procedures).

As @tommybananas mentioned:
> ...scalable applications don't handle cascades through an ORM layer.

But if it's absolutely necessary and applicable in your (and anyone interested) case, go ahead and implement a solution (as @nicoabie did, BTW good job).

To summarize, I don't think an ORM/DB should handle cascading soft deletes. But this is just my opinion.

For more details, I propose the community to do their own research on how this kind of problems are solved in existing enterprise solutions, what are the cons and pros.

I am always available for discussion ([sequelize gitter](https://gitter.im/sequelize/sequelize)).

**Comment by ephys** [comment]
Here is a feature proposal that could resolve this issue: https://github.com/sequelize/sequelize/issues/15493. Feedback welcome :)

**Comment by nicoabie** [comment]
@ephys I can't believe one year passed and I haven't finished that package.
My idea now is working on a cli that will scan an existing db and ask which delete strategy one wants to add to the different relations between tables.

We need to support the case of existing projects.

I think we could port it into sequelize once it is proven to work. I'll try to work on it more and let you know its status.

**Comment by danymajeed** [comment]
+1

**Comment by rathodajay2021** [comment]
> I have a relationship between two paranoid tables (User and Email) and I have the onDelete: 'cascade' option on the relationship ( Email.belongsTo(User, onDelete:'cascade') ).
> 
> The problem is that when I delete the user his email is **not** being deleted by cascade.
> 
> Is this a bug? Or this is how it is supposed to work?
> 
> I am using **sequelize 2.0.0-rc2** on a **postgres** database.
> 
> Thanks.
> 
> PS.: Please take a look at the code I used for test:
> 
> ```js
> var Sequelize = require('sequelize')
>     , sequelize = new Sequelize('test', 'test', 'test', {dialect: 'postgres'});
> 
> var User = sequelize.define('User', {
>     username: Sequelize.STRING,
>     birthday: Sequelize.DATE
> }, {paranoid: true}); //Without paranoid the code works fine
> 
> var Email = sequelize.define('Email', {
>     email: Sequelize.STRING,
>     primary: Sequelize.BOOLEAN
> }, {paranoid: true}); //Without paranoid the code works fine
> 
> Email.belongsTo(User, {onDelete: 'cascade'});
> 
> sequelize.sync().success(function () {
>     var userCreatedInstance = null;
> 
>     var deletedUserId = null;
>     var cascadeDeletedEmailId = null;
> 
>     User
>         .create({
>             username: 'user1',
>             birthday: new Date(1986, 06, 28)
>         })
>         .then(function (_user) {
>             userCreatedInstance = _user;
>             deletedUserId = userCreatedInstance.id;
>             return Email.create({email: 'email1@test.com', primary: true, UserId: userCreatedInstance.id});
>         })
>         .then(function (createdEmail) {
>             cascadeDeletedEmailId = createdEmail.id;
>             return userCreatedInstance.destroy();
>         })
>         .then(function () {
>             return User.find(deletedUserId);
>         })
>         .then(function (foundUser) {
>             if(foundUser){
>                 throw new Error("Shouldn't find the user with this id");
>             }
>             return Email.find(cascadeDeletedEmailId);
>         })
>         .then(function (foundEmail){
>             if(foundEmail){
>                 console.log(foundEmail.values);
>                 throw new Error("Shouldn't find the email with this id");
>             }
>         });
> });
> ```


You can overcome this issue by using afterDestroy hook.

>var User = sequelize.define('User', {
>   username: Sequelize.STRING,
>    birthday: Sequelize.DATE
>}, {paranoid: true
>hooks: {
>            afterDestroy: function (instance, options) {
>                instance.getEmail().then(email=> email.destroy()); // Softdelete on email table
>                console.log('after destroy: destroyed');
>            }
>        }
>   })
